### ISSUE TITLE
Some dashboards do not have all datasources replaced

### ISSUE THREAD

**Issue by simondeziel** [canonical/grafana-k8s-operator#178]
### Bug Description

Feeding the [LXD dashboard](https://grafana.com/grafana/dashboards/15726-lxd/) to Grafana using the `lib/charms/grafana_k8s/v0/grafana_dashboard.py` doesn't work as the JSON isn't mangled enough to replace all the "{DS_LXD}" by "{prometheusds}".

The [LXD charm](https://sdeziel.info/pub/lxd_ubuntu-22.04-amd64.charm) is the one providing the dashboard.

### To Reproduce

1. `juju deploy -m cos cos-lite --channel edge --trust --overlay ./offers-overlay.yaml --overlay ./storage-small-overlay.yaml`
2. `juju deploy -m test ./lxd_ubuntu-22.04-amd64.charm --config lxd-listen-https=true`
3. `juju integrate -m test lxd:grafana-dashboard-k8s grafana-dashboards:grafana-dashboard`

### Workaround

When saving the LXD.json dashboard into the LXD charm, one can run `sed -i 's/{DS_LXD}/{prometheusds}/g' src/grafana_dashboards/LXD.json` to workaround the issue.

### Environment

The juju controller is on a LXD container, the COS-Lite bundle is deployed in a model called `cos` running on a microk8s instance. The dashboard providing charm (LXD) runs in a different model called `test`.

### Additional context

Here's the `juju status` output of the `cos` model:

```
$ juju status -m cos --relations
Model  Controller  Cloud/Region            Version  SLA          Timestamp
cos    overlord    microk8s-cos/localhost  3.0.2    unsupported  04:36:22Z

App           Version  Status  Scale  Charm             Channel  Rev  Address         Exposed  Message
alertmanager  0.23.0   active      1  alertmanager-k8s  edge      38  10.152.183.20   no       
catalogue              active      1  catalogue-k8s     edge       6  10.152.183.211  no       
grafana       9.2.1    active      1  grafana-k8s       edge      59  10.152.183.118  no       
loki          2.4.1    active      1  loki-k8s          edge      49  10.152.183.135  no       
prometheus    2.33.5   active      1  prometheus-k8s    edge      92  10.152.183.89   no       
traefik                active      1  traefik-k8s       edge     100  172.17.33.1     no       

Unit             Workload  Agent  Address       Ports  Message
alertmanager/0*  active    idle   10.1.134.106         
catalogue/0*     active    idle   10.1.134.114         
grafana/0*       active    idle   10.1.134.99          
loki/0*          active    idle   10.1.134.74          
prometheus/0*    active    idle   10.1.134.84          
traefik/0*       active    idle   10.1.134.123         

Offer                            Application   Charm             Rev  Connected  Endpoint              Interface                Role
alertmanager-karma-dashboard     alertmanager  alertmanager-k8s  38   0/0        karma-dashboard       karma_dashboard          provider
grafana-dashboards               grafana       grafana-k8s       59   1/1        grafana-dashboard     grafana_dashboard        requirer
loki-logging                     loki          loki-k8s          49   0/0        logging               loki_push_api            provider
prometheus-receive-remote-write  prometheus    prometheus-k8s    92   0/0        receive-remote-write  prometheus_remote_write  provider
prometheus-scrape                prometheus    prometheus-k8s    92   0/0        metrics-endpoint      prometheus_scrape        requirer

Relation provider                   Requirer                     Interface              Type     Message
alertmanager:alerting               loki:alertmanager            alertmanager_dispatch  regular  
alertmanager:alerting               prometheus:alertmanager      alertmanager_dispatch  regular  
alertmanager:grafana-dashboard      grafana:grafana-dashboard    grafana_dashboard      regular  
alertmanager:grafana-source         grafana:grafana-source       grafana_datasource     regular  
alertmanager:replicas               alertmanager:replicas        alertmanager_replica   peer     
alertmanager:self-metrics-endpoint  prometheus:metrics-endpoint  prometheus_scrape      regular  
catalogue:catalogue                 alertmanager:catalogue       catalogue              regular  
catalogue:catalogue                 grafana:catalogue            catalogue              regular  
catalogue:catalogue                 prometheus:catalogue         catalogue              regular  
grafana:grafana                     grafana:grafana              grafana_peers          peer     
grafana:metrics-endpoint            prometheus:metrics-endpoint  prometheus_scrape      regular  
loki:grafana-dashboard              grafana:grafana-dashboard    grafana_dashboard      regular  
loki:grafana-source                 grafana:grafana-source       grafana_datasource     regular  
loki:metrics-endpoint               prometheus:metrics-endpoint  prometheus_scrape      regular  
prometheus:grafana-dashboard        grafana:grafana-dashboard    grafana_dashboard      regular  
prometheus:grafana-source           grafana:grafana-source       grafana_datasource     regular  
prometheus:prometheus-peers         prometheus:prometheus-peers  prometheus_peers       peer     
traefik:ingress                     alertmanager:ingress         ingress                regular  
traefik:ingress                     catalogue:ingress            ingress                regular  
traefik:ingress-per-unit            loki:ingress                 ingress_per_unit       regular  
traefik:ingress-per-unit            prometheus:ingress           ingress_per_unit       regular  
traefik:metrics-endpoint            prometheus:metrics-endpoint  prometheus_scrape      regular  
traefik:traefik-route               grafana:ingress              traefik_route          regular```

**Comment by rbarry82** [comment]
I wasn't able to reproduce this, but I looked at the dashboard JSON. You really keep up to date! It's already using a bunch of "new" Grafana 9-ish features from the export.

I have the feeling that [this PR](https://github.com/canonical/grafana-k8s-operator/pull/172) probably got it. Since some of the panels in the LXD dashboard are nested in rows; the grafana dashboard library has been mostly reverse engineered, since even the upstream "spec" -- if it weren't in jsonnet/ts/react depending on year/repo -- isn't a "clean" interface for transforming/templating libraries outside of the golden path, so we're a little limited to what we've encountered in the wild.

One of those things was panels nested inside rows, rather than rows just used as separators.

We're angling for a release today/tomorrow, so I'll re-check (or current `edge`). Current dashboard on the Grafana marketplace still the right one to reproduce, maybe?
